### PR TITLE
WIP: VERY EARLY DRAFT! WIP WIP WIP!!!! - Add clang-cl windows cross compilation

### PR DESCRIPTION
**Pull request by GregBowyer**
…

Hi, this will be a little insane and a little strange, but I would
like to get a read on the following as I refine it towards a potential
addition to these rules.

The following code allows for cross compiling windows (with MSVC ABI
rather than GNU, and I hope to adapt it to cope with clang on windows
too.

If this is something wanted I can polish these rules further rather than
fork and adapt.

This is very much an initial read, so there are some hacks, a hard-coded
path I could use help with and probably constructions that I am sure we
want to change.

As an example of this running (the binary is bytewise equivalent to the windows produced version) is as follows.

https://youtu.be/v54e1pSMUbw

### CODE REVIEW

**Comment by siddharthab** [comment]
Thanks! We would definitely welcome any Windows support in here, if we can introduce it while keeping things relatively clean. I can give you write permissions if you want. Unfortunately, I don't pay attention to this project anymore, so won't directly review your code.

File: toolchain/cc_toolchain_config.bzl [hunk 1]
```diff
@@ -358,3 +371,135 @@ def cc_toolchain_config(
...
+    msvc_env_path = "msvc_not_found"
+    msvc_env_include = "msvc_not_found"
+    msvc_env_lib = "msvc_not_found"
+    msvc_cl_path = tools_path_prefix + "clang-cl"
+    msvc_ml_path = tools_path_prefix + "clang-cl" # llvm-ml"
+    msvc_link_path = tools_path_prefix + "lld-link"
+    msvc_lib_path = tools_path_prefix + "llvm-lib"
+
+    cxx_builtin_include_directories = [
+        "/home/greg/.cache/bazel/_bazel_greg/ea13bb2946567f0877199a8bd9383c6f/" + toolchain_path_prefix + "include/c++/v1",
```
Comment (GregBowyer) [inline toolchain/cc_toolchain_config.bzl:402]: This hack is obviously wrong, it used the hermetic toolchain as downloaded, but I need to figure out how to educate bazel via skylark about how to point to the downloads without a massive hack.

File: toolchain/cc_toolchain_config.bzl [hunk 2]
```diff
@@ -358,3 +371,135 @@ def cc_toolchain_config(
...
+            sysroot_path + "/sdk/include/um",
+            sysroot_path + "/sdk/include/shared",
+        ]
+
+        lib_paths = [
+            # TODO: Arch should be inferred
+            sysroot_path + "/crt/lib/x86_64",
+            sysroot_path + "/sdk/lib/um/x86_64",
+            sysroot_path + "/sdk/lib/ucrt/x86_64",
+            # TODO: How to get this in a sysroot
```
Comment (GregBowyer) [inline toolchain/cc_toolchain_config.bzl:430]: What I mean by this TODO is that this comes from the clang bundle, but only from the windows clang bundle.

This is a bit weird as the linux clang bundle contains everything else, but omits this file, its easy to script this part in the preparation of a windows sysroot (or maybe we ask LLVM upstream to include it as part of the linux tarball).

File: toolchain/internal/common.bzl [hunk 3]
```diff
@@ -12,19 +12,23 @@
...
-SUPPORTED_TARGETS = [("linux", "x86_64"), ("linux", "aarch64"), ("darwin", "x86_64")]
+SUPPORTED_TARGETS = [("linux", "x86_64"), ("linux", "aarch64"), ("darwin", "x86_64"), ("windows", "x86_64")]
 
 host_tool_features = struct(
     SUPPORTS_ARG_FILE = "supports_arg_file",
 )
 
 toolchain_tools = [
     "clang-cpp",
+    "clang-cl",
```
Comment (GregBowyer) [inline toolchain/internal/common.bzl:23]: These are present in the linux (maybe OSX?) bundles, but I can naturally look to make them optional / stub files.

File: toolchain/tools/host_os_key.py [hunk 4]
```diff
@@ -19,7 +19,8 @@
 import sys
 
 
-_known_distros = ["freebsd", "suse", "ubuntu", "arch", "manjaro", "debian", "fedora", "centos", "amzn", "raspbian", "pop"]
+_known_distros = ["freebsd", "suse", "ubuntu", "arch", "manjaro", "debian",
+    "fedora", "centos", "amzn", "raspbian", "pop", "nixos"]
```
Comment (GregBowyer) [inline toolchain/tools/host_os_key.py:23]: Minor: This is unrelated to windows toolchains, my personal Linux is NixOS, but I am pretending it is equivalent to Ubuntu.

Will remove for a real release

File: toolchain/tools/llvm_release_name.py [hunk 5]
```diff
@@ -122,6 +122,9 @@ def _linux(llvm_version, distname, version, arch):
     elif distname == "raspbian":
         arch = "armv7a"
         os_name = "linux-gnueabihf"
+    elif distname == "nixos":
```
Comment (GregBowyer) [inline toolchain/tools/llvm_release_name.py:125]: Minor: This is unrelated to windows toolchains, my personal Linux is NixOS, but I am pretending it is equivalent to Ubuntu.

Will remove for a real release

**Comment by GregBowyer** [comment]
> Thanks! We would definitely welcome any Windows support in here, if we can introduce it while keeping things relatively clean. I can give you write permissions if you want. Unfortunately, I don't pay attention to this project anymore, so won't directly review your code.

Might be good, I will wait for others to chime in on thoughts of how to do these things.

**Comment by siddharthab** [comment]
@GregBowyer, are you still interested in pursuing this?

**Comment by siddharthab** [comment]
Closing now because of inactivity. We can reopen if needed.
